test/validation/api/ipsec/ipsec_test_in.c

```diff
@@ -1727,6 +1727,137 @@ static void test_ipsec_sa_print(void)
 	ipsec_sa_destroy(in_sa);
 }
 
+static void test_antireplay(uint32_t window_sz)
+{
+	odp_ipsec_test_sa_operation_t sa_op = ODP_IPSEC_TEST_SA_UPDATE_SEQ_NUM;
+	odp_ipsec_tunnel_param_t in_tunnel, out_tunnel;
+	odp_ipsec_sa_param_t param_in, param_out;
+	uint32_t src = IPV4ADDR(10, 0, 111, 2);
+	uint32_t dst = IPV4ADDR(10, 0, 222, 2);
+	odp_ipsec_test_sa_param_t sa_param;
+	odp_ipsec_sa_info_t sa_info;
+	ipsec_test_part test_fail_in;
+	odp_ipsec_sa_t out_sa;
+	odp_ipsec_sa_t in_sa;
+
+	memset(&in_tunnel, 0, sizeof(odp_ipsec_tunnel_param_t));
+	memset(&out_tunnel, 0, sizeof(odp_ipsec_tunnel_param_t));
+	memset(&test_fail_in, 0, sizeof(ipsec_test_part));
+
+	out_tunnel.type = ODP_IPSEC_TUNNEL_IPV4;
+	out_tunnel.ipv4.src_addr = &src;
+	out_tunnel.ipv4.dst_addr = &dst;
+
+	ipsec_sa_param_fill(&param_out,
+			    false, false, 0x4a2cbfe3, &out_tunnel,
+			    ODP_CIPHER_ALG_AES_GCM, &key_mcgrew_gcm_4,
+			    ODP_AUTH_ALG_AES_GCM, NULL,
+			    &key_mcgrew_gcm_salt_4, NULL);
+
+	ipsec_sa_param_fill(&param_in,
+			    true, false, 0x4a2cbfe3, &in_tunnel,
+			    ODP_CIPHER_ALG_AES_GCM, &key_mcgrew_gcm_4,
+			    ODP_AUTH_ALG_AES_GCM, NULL,
+			    &key_mcgrew_gcm_salt_4, NULL);
+
+	ipsec_test_part test_pass_out = {
+		.pkt_in = &pkt_mcgrew_gcm_test_4,
+		.num_pkt = 1,
+		.out = {
+			{ .status.warn.all = 0,
+			  .status.error.all = 0,
+			  .l3_type = ODP_PROTO_L3_TYPE_IPV4,
+			  .l4_type = ODP_PROTO_L4_TYPE_ESP,
+			},
+		},
+	};
+
+	ipsec_test_part test_pass_in = {
+		.num_pkt = 1,
+		.out = {
+			{ .status.warn.all = 0,
+			  .status.error.all = 0,
+			  .l3_type = ODP_PROTO_L3_TYPE_IPV4,
+			  .l4_type = ODP_PROTO_L4_TYPE_ICMPV4,
+			  .pkt_res = &pkt_mcgrew_gcm_test_4},
+		},
+	};
+
+	test_fail_in.pkt_in = &pkt_mcgrew_gcm_test_4;
+	test_fail_in.num_pkt = 1;
+	test_fail_in.out[0].status.error.antireplay = 1;
+
+	out_sa = odp_ipsec_sa_create(&param_out);
+	CU_ASSERT_NOT_EQUAL_FATAL(ODP_IPSEC_SA_INVALID, out_sa);
+
+	param_in.inbound.antireplay_ws = window_sz;
+	in_sa = odp_ipsec_sa_create(&param_in);
+	CU_ASSERT_NOT_EQUAL_FATAL(ODP_IPSEC_SA_INVALID, in_sa);
+
+	memset(&sa_info, 0, sizeof(odp_ipsec_sa_info_t));
+	CU_ASSERT_EQUAL_FATAL(0, odp_ipsec_sa_info(in_sa, &sa_info));
+
+	window_sz = sa_info.inbound.antireplay_ws;
+
+	/* 1. Advance the TOP of the window to WS * 2 */
+	sa_param.seq_num =  window_sz * 2;
+	CU_ASSERT_EQUAL(0, odp_ipsec_test_sa_update(out_sa,
```

Comment:
Use of odp_ipsec_test_sa_update() is not very nice since it is an optional part of the API and may not be supported by all implementations. Failing a validation test because odp_ipsec_test_sa_update() fails is, strictly speaking, not correct, although with current implementations it would work.

I do not think the use of odp_ipsec_test_sa_update() is necessary here. The test case could just process enough packets using the outbound SA to get the sequence number where it wants it and then after that do the actual test.

Reply:
> Use of odp_ipsec_test_sa_update() is not very nice since it is an optional part of the API and may not be supported by all implementations

We have just used this available API for updating the required seq number. Does "optional part of the API" mean, will it be deprecated soon.

> I do not think the use of odp_ipsec_test_sa_update() is necessary here. The test case could just process enough packets using the outbound SA to get the sequence number where it wants it and then after that do the actual test.

Sure, we will explore this thought.

Reply:
> We have just used this available API for updating the required seq number. Does "optional part of the API" mean, will it be deprecated soon.

Optionality means that an ODP implementation is not required to support the API. There is capability info for the sequence number update support and in addition to that the API says that the function is not required to always succeed.

Since this is API validation test code, it is not correct to fail the test (i.e. essentially declare the implementation under test as API-incompliant) just because an optional testing feature is not supported. At minimum the validation test should check the relevant capability and be prepared for the SA update function to fail. Since this would mean that the test case could not be run with all ODP implementations, it would be better if the test case did not rely on the optional feature. Tests with smallish sequence numbers do not really need the SA update feature, but potential future ESN tests probably need it.